src/server/service/kv.rs

```diff
@@ -1019,28 +1435,89 @@ impl<T: RaftStoreRouter + 'static, E: Engine, L: LockMgr> Tikv for Service<T, E,
         let storage = self.storage.clone();
         let cop = self.cop.clone();
         let gc_worker = self.gc_worker.clone();
-
-        let request_handler = stream.for_each(move |mut req| {
-            let request_ids = req.take_request_ids();
-            let requests: Vec<_> = req.take_requests().into();
-            GRPC_REQ_BATCH_COMMANDS_SIZE.observe(requests.len() as f64);
-            for (id, req) in request_ids.into_iter().zip(requests) {
-                handle_batch_commands_request(
-                    &storage,
-                    &gc_worker,
-                    &cop,
-                    &peer,
-                    id,
-                    req,
-                    tx.clone(),
+        if self.enable_req_batch {
+            let stopped = Arc::new(AtomicBool::new(false));
+            let req_batcher = ReqBatcher::new(
+                tx.clone(),
+                self.req_batch_wait_duration,
+                Arc::clone(&self.readpool_normal_thread_load),
+            );
+            let req_batcher = Arc::new(Mutex::new(req_batcher));
+            if let Some(duration) = self.req_batch_wait_duration {
+                let storage = storage.clone();
+                let req_batcher = req_batcher.clone();
+                let req_batcher2 = req_batcher.clone();
+                let stopped = Arc::clone(&stopped);
+                let start = Instant::now();
+                let timer = GLOBAL_TIMER_HANDLE.clone();
+                self.timer_pool.lock().unwrap().spawn(
+                    timer
+                        .interval(start, duration)
+                        .take_while(move |_| {
+                            // only stop timer when no more incoming and old batch is submitted.
+                            future::ok(
+                                !stopped.load(Ordering::Relaxed)
+                                    || !req_batcher2.lock().unwrap().is_empty(),
```

Comment:
Where is this result used?

Reply:
timer will keep running if we don't stop it after command stream is terminated.